CLR/Libraries/Windows_Devices/windows_devices_native_Windows_Devices_Spi_SpiDevice.cpp

```diff
@@ -0,0 +1,239 @@
+        }
+
+        GPIO_PIN misoPin = pThis[ FIELD__m_misoPin ].NumericByRef().u4;
+        if (misoPin != GPIO_PIN_NONE)
+        {
+            CPU_GPIO_ReservePin( misoPin, FALSE );
+        }
+
+        GPIO_PIN mosiPin = pThis[ FIELD__m_mosiPin ].NumericByRef().u4;
+        if (mosiPin != GPIO_PIN_NONE)
+        {
+            CPU_GPIO_ReservePin( mosiPin, FALSE );
+        }
+    }
+
+    TINYCLR_NOCLEANUP();
+}
+
+HRESULT Library_windows_devices_native_Windows_Devices_Spi_SpiDevice::TransferInternal___VOID__SZARRAY_U1__SZARRAY_U1__BOOLEAN( CLR_RT_StackFrame& stack )
+{
+    TINYCLR_HEADER();
+
+    BYTE* writeData = NULL;
+    int writeLength = 0;
+    BYTE* readData = NULL;
+    int readLength = 0;
+    CLR_RT_HeapBlock* pThis = stack.This(); FAULT_ON_NULL(pThis);
+
+    if (pThis[ FIELD__m_disposed ].NumericByRef().u1)
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_OBJECT_DISPOSED);
+    }
+
+    {
+        CLR_RT_HeapBlock_Array* writeBuffer = stack.Arg1().DereferenceArray();
+        if (writeBuffer != NULL)
+        {
+            writeData = writeBuffer->GetFirstElement();
+            writeLength = writeBuffer->m_numOfElements;
+        }
+
+        CLR_RT_HeapBlock_Array* readBuffer = stack.Arg2().DereferenceArray();
+        if (readBuffer != NULL)
+        {
+            readData = readBuffer->GetFirstElement();
+            readLength = readBuffer->m_numOfElements;
+        }
+
+        CLR_RT_HeapBlock* settings = pThis[ FIELD__m_settings ].Dereference();
+        if (settings == NULL)
+        {
+            TINYCLR_SET_AND_LEAVE(CLR_E_INVALID_OPERATION);
+        }
+
+        SPI_CONFIGURATION config = {0};
+
+        // Always disable the busy pin.
+        config.BusyPin.Pin = GPIO_PIN_NONE;
+        config.BusyPin.ActiveState = FALSE;
+
+        config.SPI_mod = pThis[ FIELD__m_spiBus ].NumericByRef().u4;
+        config.DeviceCS = settings[ SpiConnectionSettings::FIELD__m_chipSelectionLine ].NumericByRef().u4;
+        config.Clock_RateKHz = settings[ SpiConnectionSettings::FIELD__m_clockFrequency ].NumericByRef().s4 / 1000;
+
+        SpiMode mode = static_cast<SpiMode>(settings[ SpiConnectionSettings::FIELD__m_mode ].NumericByRef().s4);
+        switch (mode)
+        {
+        case SpiMode_Mode0:
+            config.MSK_IDLE = FALSE;
+            config.MSK_SampleEdge = TRUE;
+            break;
+
+        case SpiMode_Mode1:
+            config.MSK_IDLE = FALSE;
+            config.MSK_SampleEdge = FALSE;
+            break;
+
+        case SpiMode_Mode2:
+            config.MSK_IDLE = TRUE;
+            config.MSK_SampleEdge = FALSE;
+            break;
+
+        case SpiMode_Mode3:
+            config.MSK_IDLE = TRUE;
+            config.MSK_SampleEdge = TRUE;
+            break;
+        }
+
+        int dataBitLength = settings[ SpiConnectionSettings::FIELD__m_dataBitLength ].NumericByRef().s4;
+        if (dataBitLength > 8)
+        {
+            config.MD_16bits = TRUE;

```

Comment:
mmmh, MD_16bits = TRUE, but calling CPU_SPI_nWrite8_nRead8
that does not sound right, or did something changes in the driver model?